business-invariants-in-python.md

```diff
@@ -0,0 +1,269 @@
+# Enforcing business invariants in Python: type-checking, property-based testing and type combinators
+
+An application, or "program", is a set of instructions given to a computer. This program is expected to behave in a certain way, that is, closely follow a set of strict business expectations, or invariants.
+
+This article highlights the common tools we use at TCM Labs to enforce code quality in the programs that we ship to our customers. We often work with data scientists and data engineers, so examples will be given in the Python programming language.
+
+We'll be exploring a simple business domain, arithmetics, and implement a `calculus` module with an `add()` function that we expect to work on natural numbers only.
+
+## The problem
+
+At run time, ie. when the application is live and end-users are interacting with it, the Python runtime accepts calling a function with any kind of values, and not just with a subset of all values which makes sense from a business point of view.
+
+Before we go further, let's establish a bit of terminology:
+
+- at "run time" (two words) refers to when [an application is running](<https://en.wikipedia.org/wiki/Run_time_(program_lifecycle_phase)>), in contrast to [compile time](Compile_time), which is _before_ an application is running.
+- a "runtime" (single word) refers to [an execution environment](https://en.wikipedia.org/wiki/Runtime_system) (ie. Python or Node.js interpreters, a browser, etc.)
+
+In Python, avoiding bugs at run time is a responsibility which is mostly left to the programmer. More precisely, this responsability is ensuring that all functions are called with expected values, and each of them return values that make sense from a business point of view.
+
+Typically, these checks are:
+
+- not done at all within the function
+- left to the responsability of the caller
+- done in the function with a few asserts which only check for some values
+- manually done and visually tested only, using the Python REPL or Jupyter notebook
```

Comment:
Can be clearer: functions are tested with "good" values in the REPL, not the edge cases or values from another type